### PR TITLE
Add initial lexing and parsing support for `interface`s.

### PR DESCRIPTION
**Pull request by ergawy**

### CODE REVIEW

File: toolchain/parser/parser_impl.cpp [hunk 1]
```diff
@@ -634,20 +714,52 @@ auto ParseTree::Parser::ParseVariableDeclaration() -> Node {
                  /*has_error=*/!pattern || !semi);
 }
 
+auto ParseTree::Parser::ParseInterface() -> Node {
+  // TODO Error handling including the stack thing. Re-read
```
Comment (ergawy) [inline toolchain/parser/parser_impl.cpp]: Some comment clean-up is needed here and there. Sorry for the not so clean PR. Will do that after work today.

File: toolchain/parser/parser_impl.cpp [hunk 2]
```diff
@@ -556,6 +562,70 @@ auto ParseTree::Parser::ParseFunctionDeclaration() -> Node {
...
+        has_error = true;
+
+        SkipTo(tokens_.GetMatchedClosingToken(*open_bracket));
+        Consume(tokens_.GetKind(*position_));
+
+        return;
+      };
+
+      // Try to parse [me: Self].
+      auto me = ConsumeAndAddLeafNodeIf(TokenKind::Me(), ParseNodeKind::Me());
```
Comment (ergawy) [inline toolchain/parser/parser_impl.cpp]: I think we can extend `ParsePattern` to do most of the work here including for the `addr` case. Will try to do that too.

File: toolchain/lexer/token_kind_test.cpp [hunk 3]
```diff
@@ -22,7 +22,7 @@ constexpr llvm::StringLiteral SymbolRegex =
     "[\\[\\]{}!@#%^&*()/?\\\\|;:.,<>=+~-]+";
 
 // We restrict keywords to be lowercase ASCII letters and underscores.
```
Comment (zygoloid) [inline toolchain/lexer/token_kind_test.cpp]: This comment seems to be out of date now.

File: toolchain/parser/parser_impl.cpp [hunk 4]
```diff
@@ -362,6 +368,38 @@ auto ParseTree::Parser::ParsePattern(PatternKind kind) -> llvm::Optional<Node> {
                    /*has_error=*/!type);
   }
 
+  auto valid_me_param =
+      (kind == PatternKind::MeParameter && NextTokenIs(TokenKind::Me()) &&
+       tokens_.GetKind(*(position_ + 1)) == TokenKind::Colon() &&
+       tokens_.GetKind(*(position_ + 2)) == TokenKind::Self());
```
Comment (zygoloid) [inline toolchain/parser/parser_impl.cpp]: There's no requirement that `Self` be the type specified in a `me` binding nor that `Self*` be the type specified in an `addr me` binding, and we intend to allow other types to appear in that position. I think we should just be looking for `me:` or `addr me:` here, and then parse a type after that in the same way we do for the above `PatternBinding` case.

**Review by zygoloid — COMMENTED** [review]
Thanks, this generally looks good, though it's matching the expected type of `me` too narrowly.

File: toolchain/parser/parser_impl.cpp [hunk 5]
```diff
@@ -368,6 +374,36 @@ auto ParseTree::Parser::ParsePattern(PatternKind kind) -> llvm::Optional<Node> {
...
+    auto start = GetSubtreeStartPosition();
+    AddLeafNode(ParseNodeKind::Addr(), Consume(TokenKind::Addr()));
+    AddLeafNode(ParseNodeKind::Me(), Consume(TokenKind::Me()));
+    auto colon = Consume(TokenKind::Colon());
+    auto type = ParseType();
+
+    return AddNode(ParseNodeKind::PatternBinding(), colon, start,
+                   /*has_error*/ !type);
+  }
+
```
Comment (zygoloid) [inline toolchain/parser/parser_impl.cpp:404]: There seems to be a fair amount of duplication between these three code paths building a `PatternBinding`. Can you refactor this to reduce the duplication?

**Review by zygoloid — APPROVED** [review]
Thanks, looks great. If you'd prefer to land this as-is and look into refactoring the binding parsing separately, I'm fine with that.

File: toolchain/parser/parser_impl.cpp (hunk 5 above)
Comment (ergawy) [reply toolchain/parser/parser_impl.cpp:404]: Done. If you still see other chances for make the code better here, please let me know. And thanks for the review :).

File: toolchain/parser/parser_impl.cpp [hunk 6]
```diff
@@ -356,16 +361,45 @@ auto ParseTree::Parser::ParseList(TokenKind open, TokenKind close,
 
 auto ParseTree::Parser::ParsePattern(PatternKind kind) -> llvm::Optional<Node> {
   CARBON_RETURN_IF_STACK_LIMITED(llvm::None);
-  if (NextTokenIs(TokenKind::Identifier()) &&
+  auto start = GetSubtreeStartPosition();
+
+  auto continue_parsing = [&] {
```
Comment (zygoloid) [inline toolchain/parser/parser_impl.cpp]: Given that this function can handle various different kinds of pattern, I think this name isn't specific enough. How about something like `finish_parsing_pattern_binding`?

**Review by zygoloid — APPROVED** [review]

**Comment by ergawy** [comment]
Closing since it was re-implemented using the new parser.

**Pull request closed without merging**